Feature/html scrapper auxiliary

yoniabrahamy opened the pull request:
* Rendering html using Selenium on the sandbox machines and uploading them back to the cape server for processing
* Cape report will have a new section -html_scrap - which will contain information extracted from the html scrap (will currently only have urls)
* Add an auxiliary module and processing module (see diagram)
* Add geckodriver as one of the auxiliary executables we send to the virtual machine to be used by selenium

![image](https://user-images.githubusercontent.com/10047833/168466568-bac98adf-fc18-40c4-ac7b-1bd24f5990e0.png)

yoniabrahamy committed:
feat(html scraping): Implemented html scraping using selenium

yoniabrahamy committed:
feat(html-scrap): Add html scrap processing

natalyShvartzman commented on modules/processing/html_scrap.py:
```diff
@@ -0,0 +1,106 @@
+
+from lib.cuckoo.common.abstracts import Processing
+
+url_whitelist = (
```
change to allow_list

natalyShvartzman commented on analyzer/windows/modules/auxiliary/html_scrap.py:
```diff
@@ -0,0 +1,112 @@
+            log.debug('Web driver not found in path %s, not scraping', self.driver_path)
+            return
+
+        if not hasattr(self.config, "category") or self.config.category != 'file':
```
change all the " to '

davidt99 commented on analyzer/windows/modules/auxiliary/html_scrap.py:
```diff
@@ -0,0 +1,112 @@
+            opts = Options()
+            opts.headless = True
+
+            # Set fake proxy to prevent internet connectivity
```
Why do we care?

yoniabrahamy replied:
I don't want to create extra network activity which will appear in the dynamic analysis report

davidt99 replied:
but wouldn't it block some of the processing?

yoniabrahamy replied:
It should affect only the selenium execution, but I'll check it

davidt99 commented on analyzer/windows/modules/auxiliary/html_scrap.py:
```diff
@@ -0,0 +1,112 @@
+
+            sample_url = 'file:///{}'.format(os.path.abspath(file_path))
+            browser.get(sample_url)
+            time.sleep(3)
```
if possible - config it

yoniabrahamy replied:
Thing running on the VM cannot be configed, only controlled through options (which is per analysis)
I'll add an an option for it

davidt99 commented on analyzer/windows/modules/auxiliary/html_scrap.py:
```diff
@@ -0,0 +1,112 @@
+        if 'html_scrap' in options:
+            self.enabled = options['html_scrap']
+
+        self.driver_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..', 'bin',
```
can it be an absolute path?

yoniabrahamy replied:
No, these modules are copied into the machine, so we cannot trust an absolute path

davidt99 commented on analyzer/windows/modules/auxiliary/html_scrap.py:
```diff
@@ -0,0 +1,112 @@
+            self.upload_to_htmlscrap_folder('scrap.dump', browser.page_source.encode())
+
+            if not browser.current_url.startswith('file://'):
+                self.upload_to_htmlscrap_folder('scrap.dump', browser.current_url.encode())
```
it has the same name of the previous one - would it override it?

davidt99 commented on analyzer/windows/modules/auxiliary/html_scrap.py:
```diff
@@ -0,0 +1,112 @@
+            self.upload_to_htmlscrap_folder('scrap.dump', browser.page_source.encode())
+
+            if not browser.current_url.startswith('file://'):
+                self.upload_to_htmlscrap_folder('scrap.dump', browser.current_url.encode())
```
`browser.current_url.encode()` - so we would add the redirect url? if so, is there more elegant way?

yoniabrahamy replied:
Communication with the host from the guest is done by writing files, we can I guess attach it to the dump but it hurts its integrity

davidt99 commented on analyzer/windows/modules/auxiliary/html_scrap.py:
```diff
@@ -0,0 +1,112 @@
+        return True
+
+    def stop(self):
+        if self.enabled:
```
We should stop the selenuim, process here, no?

davidt99 commented on modules/processing/html_scrap.py:
```diff
@@ -0,0 +1,106 @@
+
+        html_scrap_path = os.path.join(self.analysis_path, 'htmlscrap', 'scrap.dump')
+        last_url_path = os.path.join(self.analysis_path, 'htmlscrap', 'last_url.dump')
+        if not os.path.exists(html_scrap_path):
```
isfile or it can be a directory?

davidt99 commented on modules/processing/html_scrap.py:
```diff
@@ -0,0 +1,106 @@
+        if not base64_decoded_text.isascii():
+            return None
+
+        return force_decode(base64_decoded_text.decode())
```
why the recursion? if it's needed, let's limit it to 3 level in case someone will base64 a base64

yoniabrahamy committed:
feat(html-scrap): Add html scrap processing
 - Code review fixes

yoniabrahamy committed:
feat(html-scrap): Add html scrap processing
 - Adjust browser to work with un-deprecated params
 - Prevent from gecko window to open
 - Remove network blocking, as it was blocking all FF communications
 - Extend domain whitelist due to additional network comms
